### DISCUSSION TITLE
Ability to capture a regex and use it later

### DISCUSSION THREAD

**Discussion by mehulkar** [prysk/prysk#160] [Ideas]
I would like to be able to use regex to capture part of the output of a command and then use it later. I'm not sure of the syntax, but the general use case is:

1. Run an command that generates some files with non-deterministic names
2. The non-deterministic names are printed to stdout
3. Capture the filenames
4. Run `cat` on those files and assert against the contents printed in output

**Comment by mehulkar** [comment]
My workaround right now is to redirect the output of (1) to a file, `cat` it and in then in a second step `cat filename | grep ...`

**Comment by nigels-com** [comment]
Non-determinism does tend to add complexity to testing.
Is the non-determinism necessary, actually?

**Comment by Nicoretti** [comment]
Hi @mehulkar,

I think I would have used the same approach (`"workaround"`) for such a scenario.
Even though I am not sure if we should add such a feature I  am curious, how  the ideal syntax for this would look like?
Could you write an example test file which is using such a feature?

**Comment by mehulkar** [comment]
> Is the non-determinism necessary, actually?

It's sometimes hard to workaround. We are using this in Turborepo ([`vercel/turbo`](//github.com/vercel/turbo)), which is a task runner / build system. One of the main features is that tasks that don't depend on each other can run in parallel, so the order of the output can change between runs.

> how the ideal syntax for this would look like?

That's a good question. Maybe saving named capture groups as env variables would work?

```
  $ turbo run build
  Output file: (?<something>[\w]+)\.txt (re)

  $ cat "$something.txt"
  my output
```

**Comment by nigels-com** [comment]
One other trick I sometimes use is to `| sort` the output to put the lines into a deterministic order.

**Comment by Nicoretti** [comment]
Hi @mehulkar,


I was doing some thinking regarding the feature you proposed. Even though I understand the beauty and convenience 
from an enduser point of view, I am not convinced that capturing and providing that kind of context should be part of `prysk` itself (execution engine, syntax, etc.). 

     ℹ️  If you are interested in the "why", e.g. because you want to address the concerns I have, let me know.

So I was wondering what makes your workaround "bad", or at least why you want to get rid of it?
(Feel free to extend and correct the points/assumptions bellow)

* Not explicit enough for the reader
* A very custom approach for a general problem


Thinking about all of that I was wondering if it would be a legit solution to provide a command line tool (helper) which ships with the `prysk` package whose purpose explicitly addresses the context issue.

E.g.:

```
$ turbo run build | prysk-context put KEYNAME --regex "Output file: (?<something>[\w]+)\.txt"

$ cat $(prysk-context get KEYNAME)
the output
```

**What do you think?**

### 🗒️  Side Note:

   Over time prysk could even ship a set of its own primitives `prysk-ls, prysk-cat, ...`,
   so the tests can get more independent form the provided platform/environment.

**Comment by nigels-com** [comment]
I've thought for a while that Python itself ought to be a good basis for a portable set of CLI tools implementing the same functionality as commonly used CLI tools.  grep, sort, cat, and so on.  The side note about a potential `prysk-ls` and so does seem to dovetail with that - very much interested in using prysk in a more cross-platform way.  Just to point out that `cmake` [does include some commonly needed tools](https://cmake.org/cmake/help/latest/manual/cmake.1.html#run-a-command-line-tool) (such as tar) in a way that is very helpful for cross-platform builds.
